types/audit-details.d.ts

```diff
@@ -5,63 +5,166 @@
  */
 
 declare global {
-  module LH.Audit.Details {
-    export interface Filmstrip {
-      type: 'filmstrip';
-      scale: number;
-      items: {
-        /** The relative time from navigationStart to this frame, in milliseconds. */
-        timing: number;
-        /** The raw timestamp of this frame, in microseconds. */
+  module LH.Audit {
+    export type Details =
+      Details.CriticalRequestChain |
+      Details.Diagnostic |
+      Details.Filmstrip |
+      Details.Opportunity |
+      Details.Screenshot |
+      Details.Table;
+
+    // Details namespace.
+    export module Details {
+      export interface CriticalRequestChain {
+        type: 'criticalrequestchain';
+        longestChain: {
+          duration: number;
+          length: number;
+          transferSize: number;
+        };
+        chains: Audit.SimpleCriticalRequestNode;
+      }
+
+      export interface Filmstrip {
+        type: 'filmstrip';
+        scale: number;
+        items: {
+          /** The relative time from navigationStart to this frame, in milliseconds. */
+          timing: number;
+          /** The raw timestamp of this frame, in microseconds. */
+          timestamp: number;
+          /** The data URL encoding of this frame. */
+          data: string;
+        }[];
+      }
+
+      export interface Opportunity {
+        type: 'opportunity';
+        overallSavingsMs: number;
+        overallSavingsBytes?: number;
+        headings: OpportunityColumnHeading[];
+        items: OpportunityItem[];
+        diagnostic?: Diagnostic;
+      }
+
+      export interface Screenshot {
+        type: 'screenshot';
         timestamp: number;
-        /** The data URL encoding of this frame. */
         data: string;
-      }[];
-    }
+      }
 
-    export interface Screenshot {
-      type: 'screenshot';
-      timestamp: number;
-      data: string;
-    }
+      export interface Table {
+        type: 'table';
+        headings: TableColumnHeading[];
+        items: TableItem[];
+        summary?: {
+          wastedMs?: number;
+          wastedBytes?: number;
+        };
+        diagnostic?: Diagnostic;
```

Comment:
turns out I couldn't do `[p: string]: Diagnostic` here because that means *any* string accessor of `Table` should return a value assignable to `Diagnostic`, which we don't want. So this is a middle ground, and seems flexible enough (@exterkamp rightly pointed out that if a `Diagnostic` object can contain anything, why would you ever need two of them as properties on the same object?)

Reply:
being able to define a set of properties with certain types and a different type for every other property not explicitly defined will be possible if/when negated types land in tsc (https://github.com/Microsoft/TypeScript/pull/29317)